api/v1alpha1/rollouthistory_types.go

```diff
@@ -0,0 +1,163 @@
+	TrafficRouting TrafficRoutingInfo `json:"trafficRouting,omitempty"`
+}
+
+type NameAndSpecData struct {
+	// Name indicates the name of object ref, such as rollout name, workload name, ingress name, etc.
+	Name string `json:"name"`
+	// Data indecates the spec of object ref
+	// +kubebuilder:pruning:PreserveUnknownFields
+	// +kubebuilder:validation:Schemaless
+	Data runtime.RawExtension `json:"data,omitempty"`
+}
+
+// RolloutInfo indicates information of the rollout related
+type RolloutInfo struct {
+	// RolloutID indicates the new rollout
+	// if there is no new RolloutID this time, ignore it and not execute RolloutHistory
+	RolloutID       string `json:"rolloutID"`
+	NameAndSpecData `json:",inline"`
+}
+
+// ServiceInfo indicates information of the service related
+type ServiceInfo struct {
+	NameAndSpecData `json:",inline"`
+}
+
+// TrafficRoutingInfo indicates information of Gateway API or Ingress
+type TrafficRoutingInfo struct {
+	// IngressRef indicates information of ingress
+	// +optional
+	Ingress *IngressInfo `json:"ingress,omitempty"`
+	// HTTPRouteRef indacates information of Gateway API
+	// +optional
+	HTTPRoute *HTTPRouteInfo `json:"httpRoute,omitempty"`
+}
+
+// IngressInfo indicates information of the ingress related
+type IngressInfo struct {
+	NameAndSpecData `json:",inline"`
+}
+
+// HTTPRouteInfo indicates information of gateway API
+type HTTPRouteInfo struct {
+	NameAndSpecData `json:",inline"`
+}
+
+// WorkloadInfo indicates information of the workload, such as cloneset, deployment, advanced statefulset
+type WorkloadInfo struct {
+	metav1.TypeMeta `json:",inline"`
+	NameAndSpecData `json:",inline"`
+}
+
+// RolloutHistoryStatus defines the observed state of RolloutHistory
+type RolloutHistoryStatus struct {
+	// INSERT ADDITIONAL STATUS FIELD - define observed state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+
+	// Phase indicates phase of RolloutHistory, such as "pending", "updated", "completed"
+	Phase string `json:"phase,omitempty"`
+	// RolloutState indicates the rollouts status
+	RolloutState RolloutState `json:"rolloutState,omitempty"`
+	// CanarySteps indicates the pods released each step
+	CanarySteps []CanaryStepInfo `json:"canarySteps,omitempty"`
+}
+
+// RolloutState indicates the rollouts status
+type RolloutState struct {
+	// RolloutPhase is the rollout phase.
+	RolloutPhase RolloutPhase `json:"phase,omitempty"`
+	// Message provides details on why the rollout is in its current phase
+	Message string `json:"message,omitempty"`
+}
+
+// CanaryStepInfo indicates the pods for a revision
+type CanaryStepInfo struct {
+	// CanaryStepIndex indicates step this revision
+	CanaryStepIndex int32 `json:"canaryStepIndex,omitempty"`
+	// Pods indicates the pods information
+	Pods []Pod `json:"pods,omitempty"`
+}
+
+// Pod indicates the information of a pod, including name, ip, node_name.
+type Pod struct {
```

Comment:
Do we need to record whether this Pod is updated successfully or not? cc @zmberg

Reply:
In rollout-history controller, it will record pod information after `rollout.status...canaryStepState` has been CanaryStepPaused  where this step pods have been released . and I think this pods should be successfully if they are recorded. Thanks!